### PR TITLE
Add OpenAICompatClient for OpenAI-compatible endpoints

### PR DESCRIPTION
**Pull request by lucasgerads**
Tested end-to-end against Cloudflare Workers AI (Mistral-Small-3.1-24B) via my application (teasel local). Multi-turn function calling, streaming, bearer auth all confirmed working. 

Provider-agnostic client for any backend exposing the OpenAI /v1/chat/completions shape with optional bearer auth — Cloudflare Workers AI, Fireworks, OpenRouter, OpenAI itself, etc. Native function calling, SSE streaming, full LLMClient protocol surface.

Addresses the review feedback in #88:
- Mirrors VLLMClient sampling kwargs (top_k, min_p, repeat_penalty, presence_penalty, chat_template_kwargs) with recommended_sampling defaulting to False (hosted model IDs aren't in forge's registry).
- Accepts passthrough and inbound_anthropic_body to satisfy the LLMClient protocol; inbound_anthropic_body is a no-op here.
- extra_headers kwarg for provider quirks (e.g. OpenRouter's HTTP-Referer) without a per-provider registry.
- aclose() for httpx pool cleanup.
- Docstring sets the contributor expectation: file an issue rather than adding per-provider if/else branches.

Includes BACKEND_SETUP.md section and CHANGELOG entry.

### CODE REVIEW

File: src/forge/clients/openai_compat.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    def _parse_tool_calls(tool_calls: list[dict[str, Any]]) -> list[ToolCall]:
+        """Parse OpenAI tool_calls — arguments are JSON strings."""
+        parsed: list[ToolCall] = []
+        for tc in tool_calls:
+            fn = tc.get("function", {})
+            raw_args = fn.get("arguments") or "{}"
+            try:
+                args = json.loads(raw_args) if isinstance(raw_args, str) else raw_args
+            except json.JSONDecodeError:
+                args = {}
```
Comment (antoinezambelli) [inline src/forge/clients/openai_compat.py]: Could you change this to match `LlamafileClient`'s malformed-args behavior rather than silently coercing to `{}`?

Forge's intended behavior here is: malformed native tool-call JSON should route through the normal retry/nudge path, not become an executable empty-args tool call. Otherwise a provider/model can emit invalid arguments and Forge may still proceed with fn(**{}), which is exactly the kind of quiet false success the library tries to avoid.

The precedent is `LlamafileClient`, which returns a `TextResponse` on a `JSONDecodeError` so the response routes back into the rescue-parse + retry loop:
- non-streaming: src/forge/clients/llamafile.py:523-528
- streaming: src/forge/clients/llamafile.py:400-414

Note: `VLLMClient` currently raises instead of retrying in the same case — that's existing debt on my side, not the behavior I want this new client to copy. I'll patch it up after this lands.

File: src/forge/clients/openai_compat.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                    )
+                    fn = tc.get("function", {})
+                    if fn.get("name"):
+                        slot["function"]["name"] += str(fn["name"])
+                    # OpenAI streaming sends `arguments` as string fragments
+                    # that we concatenate into the final JSON string. Anything
+                    # non-string would be a non-compliant provider; drop it
+                    # rather than json.dumps a partial object into the buffer.
+                    args_frag = fn.get("arguments")
+                    if isinstance(args_frag, str):
```
Comment (antoinezambelli) [inline src/forge/clients/openai_compat.py]: Same root issue as the non-streaming case above, in two places. Right now a non-string arguments fragment is silently skipped while assembly continues from later fragments — so a single bad fragment produces a tool call with a gap in its arguments JSON, which then fails to parse and falls into the same {} path as `_parse_tool_calls`.

Preferred fix is the same as comment 1: malformed/incomplete streamed tool args should end as a `TextResponse` final result so Forge's validator can drive the retry/nudge loop, matching `LlamafileClient`'s streaming path (src/forge/clients/llamafile.py:400-414). Treat it as one fix in two spots rather than two separate asks.

**Review by antoinezambelli — CHANGES_REQUESTED** [review]
Thanks for this — the overall shape is spot on and it matches the direction from #88: standalone sibling client, provider-agnostic OpenAI-compatible endpoint, bearer auth, extra_headers, full LLMClient surface, and good focused tests/docs.

One change before merge, around Forge's fail-fast/fail-loud preference: malformed tool-call argument JSON shouldn't become {}. Please route it through the retry-driving TextResponse path, matching `LlamafileClient` — both in _parse_tool_calls (non-streaming) and in the streaming assembly. The two inline comments point at both spots; they're the same fix in two places.

Note: `VLLMClient` currently doesn't fully match the desired malformed-args behavior either. That's existing debt on my side, not precedent to copy here — `LlamafileClient` is the reference.

**Comment by lucasgerads** [comment]
Addressed both comments with the single fix you described: malformed tool-call args now return a `TextResponse` (via `_parse_tool_calls`, shared by the streaming + non-streaming paths) instead of coercing to `{}`, matching `LlamafileClient`. Also stopped the streaming path from silently dropping a non-string arg fragment, and added test coverage. Left `VLLMClient` alone per your note.

Heads-up: I rebased on `main` and force-pushed, so your inline comments show as outdated — the fix is in the latest commit (`660a53c`).

**Review by antoinezambelli — APPROVED** [review]
This is great — thank you! Merging it.

The fail-loud fix is exactly right, and I really like that you mirrored `LlamafileClient` — that's the correct reference here (not `VLLMClient`), and it keeps the two clients behaving consistently:

- **Non-streaming** (`_parse_tool_calls`): returning a `TextResponse` on `JSONDecodeError` instead of coercing to `fn(**{})` is precisely the behavior we want — it lets the validator's rescue-parse + retry/nudge loop drive a correction rather than quietly succeeding with empty args. Surfacing assistant text when present and falling back to the raw args otherwise is a nice touch. Matches `llamafile.py:524-528`.
- **Streaming** (`send_stream`): nice catch routing final assembly through the same path, and serializing a non-string arg fragment instead of dropping it is the right instinct — a dropped fragment can leave a gap that parses into wrong-but-valid args (a quiet false success), whereas folding it in means the parse at stream end either recovers the object or fails loud into the retry path. Matches `llamafile.py:411-414`.

The "one malformed among several → whole response becomes `TextResponse`" behavior (no partial execution of valid sibling calls) lines up with llamafile's bail too, and the tests around it are lovely. The `extra_headers`-over-bearer design keeping providers out of a per-quirk registry is exactly the boundary I'd have wanted.

I pulled it down and ran everything locally: your 38 new tests pass, and the full unit suite (1066 tests) is green against current `main`.

Really appreciate how thoroughly you turned around the review feedback — the malformed/streaming/dict-args edge-case tests especially. Squash-merging now. Thanks again for the contribution!

**Comment by lucasgerads** [comment]
Thanks for merging and the thorough review!
